tests/test_cases/test_axi4/test_axi4.py

```diff
@@ -0,0 +1,451 @@
+        if written != read:
+            raise TestFailure("#{}: wrote {:#x} but read back {:#x}"
+                              .format(i, written, read.integer))
+
+
+@cocotb.test()
+async def test_axi4lite_write_burst(dut):
+    """Test that write bursts are correctly refused by the AXI4-Lite driver"""
+
+    axim = AXI4LiteMaster(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, _ = get_parameters(dut)
+    length = randint(2, 16)
+
+    await setup_dut(dut)
+
+    try:
+        await axim.write(ram_start, [randrange(0, 2**(data_width * 8))
+                                     for i in range(length)])
+
+        raise TestFailure(
+            "Write with length={} has been performed by the driver, but "
+            "burst operations are not allowed on AXI4-Lite".format(length))
+
+    except ValueError:
+        pass
+
+
+@cocotb.test()
+async def test_read_length_mismatch(dut):
+    """Test that a mismatch in the read data length is correctly identified"""
+
+    class AXI4Master_unmanaged_arlen(AXI4Master):
+        _signals = [signal for signal in AXI4Master._signals
+                    if signal != "ARLEN"]
+
+    axim = AXI4Master_unmanaged_arlen(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, ram_stop = get_parameters(dut)
+    length = randint(2, 16)
+
+    burst_length = randint(2, 255)
+    base = randrange(ram_start, ram_stop, 4096)
+    offset = randrange(0, 4096 - (burst_length - 1) * data_width, data_width)
+    address = base + offset
+
+    await setup_dut(dut)
+
+    for length_delta in (-1, 1):
+        try:
+            # Override the driver's ARLEN value, forcing a wrong one
+            await RisingEdge(dut.clk)
+            arlen = burst_length - 1 + length_delta
+            getattr(dut, AXI_PREFIX + '_ARLEN') <= arlen
+            await axim.read(address, burst_length)
+
+            raise TestFailure("Mismatch between ARLEN value ({}) and number "
+                              "of read words ({}), but the driver did not "
+                              "raise an exception"
+                              .format(arlen, burst_length))
+
+        except AXIReadBurstLengthMismatch:
+            pass
+
+
+single_beat = TestFactory(test_single_beat)
+single_beat.add_option('driver', (AXI4Master, AXI4LiteMaster))
+single_beat.add_option('address_latency', (0,))
+single_beat.add_option('data_latency', (0,))
+single_beat.generate_tests()
+
+single_beat_with_latency = TestFactory(test_single_beat)
+single_beat_with_latency.add_option('driver', (AXI4Master,))
+single_beat_with_latency.add_option('address_latency', (0, 5))
+single_beat_with_latency.add_option('data_latency', (1, 10))
+single_beat_with_latency.generate_tests(postfix="_latency")
+
+incr_burst = TestFactory(test_incr_burst)
+incr_burst.add_option('return_rresp', (True, False))
+incr_burst.generate_tests()
+
+# AXI4Master cannot be tested, as axi_interconnect.v acknowledges only one
+# WDATA for unmapped addresses in some cases
```

Comment:
If this is a bug in the interconnect, then it should be tracked down and fixed.  If you can provide more info, that would be great!

Reply:
You can reproduce the bug with:

```
@cocotb.test()
async def test_bug(dut):
    axim = AXI4Master(dut, AXI_PREFIX, dut.clk)
    ram_stop = get_parameters(dut)[3]

    await setup_dut(dut)
    await First(ClockCycles(dut.clk, 32), axim.write(ram_stop, [0x00000000] * 4))
```

This is the result:
![bug](https://user-images.githubusercontent.com/6459604/96404849-3a171280-11dc-11eb-826e-137f1976a863.png)

Let me know if you need more details.

Reply:
Yep, that's a bug alright.  In STATE_WRITE_DROP, please try ANDing in wlast like so:

    if (current_s_axi_wready && current_s_axi_wvalid && current_s_axi_wlast) begin

and let me know if that fixes the problem.

Reply:
Yes, it fixes it:
![bug](https://user-images.githubusercontent.com/6459604/96411176-56b94780-11e8-11eb-979f-74c2be0112e4.png)

Reply:
Excellent; that fix is now in the upstream repo.